### PR TITLE
Add support for union and intersection types

### LINKED ISSUE #643: Scoper is not able to scope union return types
## Bug report

| Question    | Answer
| ------------| ---------------
| PHP-Scoper version | 0.16.2
| PHP version | 8.1
| Platform with version | MacOS
| Github Repo | - 

Given following class:
https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Config/Definition/Builder/TreeBuilder.php#L35-L38

Actual
```php
public function getRootNode() : \Symfony\Component\Config\Definition\Builder\NodeDefinition|\Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition
```

Expected:
```php
public function getRootNode() : \_PhpScoperae3dc8cfa6ce\Symfony\Component\Config\Definition\Builder\NodeDefinition|\_PhpScoperae3dc8cfa6ce\Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition
```

### PR DESCRIPTION
**Pull request by veewee**
Fixes #643

### CODE REVIEW

File: specs/union.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    
+    class X
+    {
+        public function foo(Y|Z $a, null|Y $b) : null|Y|Z
+        {
+        }
+    }
+    
+    PHP,
+    'Function casts' => <<<'PHP'
```
Comment (theofidry) [inline specs/union.php:61]: I would suggest adding a blank line in-between

File: specs/union.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    
+    ----
+    <?php
+    
+    namespace Humbug;
+    
+    class X
+    {
+        public function foo(Y|Z $a, null|Y $b) : null|Y|Z
```
Comment (theofidry) [inline specs/union.php:55]: might be worth adding `static` or `self` into the mix as well? Isn't there a nullable variant too?

If this is too bloated for a single method it's also find to have those cases within different functions:

```
        public function method0(Y|Z $a) : Y|Z
        public function method1(Y|null $a) : Y|null
// ...
```

**Review by theofidry — COMMENTED** [review]
Thanks a lot! I'm really be happy if that's the only change necessary for adding union support 😄

File: specs/union.php (hunk 2 above)
Comment (veewee) [reply specs/union.php]: Added these cases as well and covered traits and interfaces as well.

**Comment by veewee** [comment]
> Thanks a lot! I'm really be happy if that's the only change necessary for adding union support 😄

Yeah me too! Looks like the system is set up in a very flexible way. Good job!
I also took the liberty of adding support for intersection types. Or would you rather have a separate PR for that once this one is done?

We already created a version of grumphp that uses the union types with box, that seems to work ok.

**Comment by theofidry** [comment]
Thank you @veewee

**Pull request merged**